NEXT-12776 -  [MIGRATE] Administration: Using mixins (#175)
shopware/docs: guides/plugins/plugins/administration/using-mixins.md

# Using Mixins

## Overview

This documentation chapter will cover how to use an existing administration Mixin in your plugin.
Generally, mixins behave the same as they do in Vue normally, differing only in the registration and the way mixins are included in a component.

## Prerequisites

All you need for this guide is a running Shopware 6 instance and full access to both the files and a running plugin.
Of course you'll have to understand JavaScript, but that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation.
As stated before mixins in Shopware are basically the same as in Vue, so you should have read their [documentation](https://vuejs.org/v2/guide/mixins.html) on them first. 

## Finding a mixin

The Shopware 6 Administration comes with a few predefined mixins, you can find a list of all of them [here](./../../../../resources/references/core-reference/administration-reference/mixins.md)

> If you want to learn how to create your own mixn look at this guide: [Creating Mixins](./add-mixins.md)

Reviewer: ```suggestion
If you want to learn how to create your own mixn look at this guide: [Creating mixins](./add-mixins.md)
```

## Using the Mixin

After we've found the mixin we need,
we can get it from the registry with the `Mixin.getByName` function and inject it into our component as seen below.

> In this example we'll use the notification mixin.

Author: i can't think of anything better so i guess i'll apply this

{% code title="<administration root>/components/swag-basic-example/index.js" %}
```javascript
const { Component, Mixin } = Shopware;

Component.register('swag-basic-example', {

    mixins: [
        Mixin.getByName('notification')
    ],

    methods: {

>         greet: function (){

Reviewer: ```suggestion
        greet: function () {
```

            this.createNotificationSuccess({ title: 'Greetings' })
        }
    }
});
```
{% endcode %}

## Next steps

> Now that you learned how to use a mixin, you might want to create or customize an administration component or create your own mixn:

Reviewer: ```suggestion
Now that you learned how to use a mixin, you might want to create or customize an administration component or create your own mixin:
```

* [Creating a new administration component](./add-custom-component.md)
* [Extending an existing administration component](./customizing-components.md)

> * [Creating a Mixin](./add-mixins.md)

Reviewer: ```suggestion
* [Creating a mixin](./add-mixins.md)
```

# Using Mixins

## Overview

This documentation chapter will cover how to use an existing administration Mixin in your plugin.
Generally, mixins behave the same as they do in Vue normally, differing only in the registration and the way mixins are included in a component.

## Prerequisites

All you need for this guide is a running Shopware 6 instance and full access to both the files and a running plugin.
Of course you'll have to understand JavaScript, but that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation.
As stated before mixins in Shopware are basically the same as in Vue, so you should have read their [documentation](https://vuejs.org/v2/guide/mixins.html) on them first. 

> ## Finding a Mixin

Reviewer 2: ```suggestion
## Finding a mixin
```

The Shopware 6 Administration comes with a few predefined mixins, here is a list of all of them:

> | Mixin                       | Description                                                                                                                                 | Link                                                                                                                                                          |

Reviewer 2: Wouldn't that fit better as a reference page? Even if it's a small one :)

|-----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | discard-detail-page-changes | Mixin which resets entity changes on page leave or if the id of the entity changes. This also affects changes in associations of the entity | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/discard-detail-page-changes.mixin.js) |

Reviewer 2: Mixin names in code syntax please, e.g.:

```suggestion
| `discard-detail-page-changes` | Mixin which resets entity changes on page leave or if the id of the entity changes. This also affects changes in associations of the entity | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/discard-detail-page-changes.mixin.js) |
```

| sw-form-field               | This mixin is used to provide common functionality between form fields                                                                      | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/form-field.mixin.js)                  |

> | listing                     | Mixin wich is used in almost all listing pages to for example keep track of the current page of the administration                          | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/listing.mixin.js)                     |

Reviewer 2: ```suggestion
| listing                     | Mixin which is used in almost all listing pages to for example keep track of the current page of the administration                          | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/listing.mixin.js)                     |
```

> | notification                | This mixin is used to easier create notifications in the administrations                                                                    | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/notification.mixin.js)                |

Reviewer 2: ```suggestion
| notification                | This mixin is used to create notifications in the administrations more easily                                                                   | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/notification.mixin.js)                |
```

| placeholder                 | Provides a function to localize placeholders                                                                                                | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/placeholder.mixin.js)                 |
| position                    | A Mixin which contains helpers to work with position integers                                                                               | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/position.mixin.js)                    |

> | remove-api-error            | This mixin removes API errors after the user for example corrected a invalid input i.e. leaving the product name field blank                | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/remove-api-error.mixin.js)            |

Reviewer 2: ```suggestion
| remove-api-error            | This mixin removes API errors e.g. after the user corrected a invalid input i.e. leaving the product name field blank                | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/remove-api-error.mixin.js)            |
```

| salutation                  | A common adapter for the `salutation` filter                                                                                                | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/salutation.mixin.js)                  |
| sw-inline-snippet           | Makes it possible to use snippets inline                                                                                                    | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/sw-inline-snippet.mixin.js)           |
| validation                  | Is used to validate inputs in various form fields                                                                                           | [link](https://github.com/shopware/platform/blob/v6.3.5.1/src/Administration/Resources/app/administration/src/app/mixin/validation.mixin.js)                  |

If you want to learn how to create your own mixn look at this guide: [Creating Mixins](./add-mixins.md)

## Using the Mixin

> After we've found the mixin we need.

Reviewer 2: ```suggestion
After we've found the mixin we need,
```

> We can get it from the registry with the `Mixin.getByName` function and inject it into our component as seen below.

Reviewer 2: ```suggestion
we can get it from the registry with the `Mixin.getByName` function and inject it into our component as seen below.
```

In this example we'll use the notification mixin.

{% code title="<administration root>/components/swag-basic-example/index.js" %}
```javascript
const { Component, Mixin } = Shopware;

Component.register('swag-basic-example', {

    mixins: [
        Mixin.getByName('notification')
    ],

    methods: {
        greet: function (){
            this.createNotificationSuccess({ title: 'Greetings' })
        }
    }
});
```
{% endcode %}

## Next steps

Now that you learned how to use a mixin, you might want to create or customize an administration component or create your own mixn:
* [Creating a new administration component](./add-custom-component.md)
* [Extending an existing administration component](./customizing-components.md)
* [Creating a Mixin](./add-mixins.md)

> If you however want to add your own service to the administration:

Reviewer 2: ```suggestion
However, if you want to add your own service to the administration take a look at this guide:
```
